IndexOutOfBoundsException while compiling INSERT_STMT with subexpressions

Library version: `1.1.3`

The schema and query described in sq file below

```
CREATE TABLE category(
  id TEXT NOT NULL PRIMARY KEY,
  name TEXT NOT NULL,
  description TEXT NOT NULL
);

save:
INSERT OR REPLACE
INTO category (rowid, id, name, description)
VALUES (COALESCE((SELECT rowid FROM category WHERE id = ?1), NULL), ?1, ?2, ?3);
```

Result in the following exception during a gradle build:
```
Caused by: java.lang.IllegalStateException: Failed to compile SqlDelightInsertStmtImpl(INSERT_STMT): [] :
INSERT OR REPLACE
INTO category (rowid, id, name, description)
VALUES (COALESCE((SELECT rowid FROM category WHERE id = ?1), NULL), ?1, ?2, ?3)
	at com.squareup.sqldelight.core.compiler.SqlDelightCompilerKt.tryWithElement(SqlDelightCompiler.kt:199)
	at com.squareup.sqldelight.core.compiler.QueriesTypeGenerator.addExecute(QueriesTypeGenerator.kt:123)
	at com.squareup.sqldelight.core.compiler.QueriesTypeGenerator.interfaceType(QueriesTypeGenerator.kt:48)
	at com.squareup.sqldelight.core.compiler.SqlDelightCompiler.writeQueriesInterface$sqldelight_compiler(SqlDelightCompiler.kt:149)
	at com.squareup.sqldelight.core.compiler.SqlDelightCompiler.writeInterfaces(SqlDelightCompiler.kt:43)
	at com.squareup.sqldelight.core.SqlDelightEnvironment$generateSqlDelightFiles$2.invoke(SqlDelightEnvironment.kt:113)
	at com.squareup.sqldelight.core.SqlDelightEnvironment$generateSqlDelightFiles$2.invoke(SqlDelightEnvironment.kt:53)
	at com.alecstrong.sqlite.psi.core.SqliteCoreEnvironment$forSourceFiles$1.processFile(SqliteCoreEnvironment.kt:86)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:117)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:114)
	at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContent(SqliteCoreEnvironment.kt:108)
	at com.alecstrong.sqlite.psi.core.SqliteCoreEnvironment.forSourceFiles(SqliteCoreEnvironment.kt:84)
	at com.squareup.sqldelight.core.SqlDelightEnvironment.generateSqlDelightFiles(SqlDelightEnvironment.kt:110)
	at com.squareup.sqldelight.gradle.SqlDelightTask.generateSqlDelightFiles(SqlDelightTask.kt:58)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:48)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:41)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:705)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:672)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$4.run(ExecuteActionsTaskExecuter.java:338)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:327)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:312)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:75)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:158)
	at org.gradle.internal.execution.impl.steps.ExecuteStep.execute(ExecuteStep.java:46)
	at org.gradle.internal.execution.impl.steps.CancelExecutionStep.execute(CancelExecutionStep.java:34)
	at org.gradle.internal.execution.impl.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:69)
	at org.gradle.internal.execution.impl.steps.TimeoutStep.execute(TimeoutStep.java:49)
	at org.gradle.internal.execution.impl.steps.CatchExceptionStep.execute(CatchExceptionStep.java:34)
	at org.gradle.internal.execution.impl.steps.CreateOutputsStep.execute(CreateOutputsStep.java:49)
	at org.gradle.internal.execution.impl.steps.SnapshotOutputStep.execute(SnapshotOutputStep.java:42)
	at org.gradle.internal.execution.impl.steps.SnapshotOutputStep.execute(SnapshotOutputStep.java:28)
	at org.gradle.internal.execution.impl.steps.CacheStep.executeWithoutCache(CacheStep.java:133)
	at org.gradle.internal.execution.impl.steps.CacheStep.lambda$execute$5(CacheStep.java:83)
	at org.gradle.internal.execution.impl.steps.CacheStep.execute(CacheStep.java:82)
	at org.gradle.internal.execution.impl.steps.CacheStep.execute(CacheStep.java:37)
	at org.gradle.internal.execution.impl.steps.PrepareCachingStep.execute(PrepareCachingStep.java:33)
	at org.gradle.internal.execution.impl.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:38)
	at org.gradle.internal.execution.impl.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:23)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:95)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:88)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:52)
	at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:36)
	at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:34)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:109)
	... 129 more
Caused by: java.lang.IndexOutOfBoundsException: Index: 3, Size: 3
	at com.squareup.sqldelight.core.lang.util.ArgumentsKt.argumentType(Arguments.kt:126)
	at com.squareup.sqldelight.core.lang.util.ArgumentsKt.inferredType(Arguments.kt:77)
	at com.squareup.sqldelight.core.lang.util.ArgumentsKt.argumentType(Arguments.kt:58)
	at com.squareup.sqldelight.core.compiler.model.BindableQuery$arguments$2.invoke(BindableQuery.kt:85)
	at com.squareup.sqldelight.core.compiler.model.BindableQuery$arguments$2.invoke(BindableQuery.kt:35)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at com.squareup.sqldelight.core.compiler.model.BindableQuery.getArguments$sqldelight_compiler(BindableQuery.kt)
	at com.squareup.sqldelight.core.compiler.model.BindableQuery$parameters$2.invoke(BindableQuery.kt:55)
	at com.squareup.sqldelight.core.compiler.model.BindableQuery$parameters$2.invoke(BindableQuery.kt:35)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at com.squareup.sqldelight.core.compiler.model.BindableQuery.getParameters$sqldelight_compiler(BindableQuery.kt)
	at com.squareup.sqldelight.core.compiler.ExecuteQueryGenerator.interfaceFunction(ExecuteQueryGenerator.kt:27)
	at com.squareup.sqldelight.core.compiler.QueriesTypeGenerator$addExecute$1.invoke(QueriesTypeGenerator.kt:131)
	at com.squareup.sqldelight.core.compiler.QueriesTypeGenerator$addExecute$1.invoke(QueriesTypeGenerator.kt:22)
	at com.squareup.sqldelight.core.compiler.SqlDelightCompilerKt.tryWithElement(SqlDelightCompiler.kt:197)
	... 188 more
```

--- a/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/util/Arguments.kt
+++ b/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/util/Arguments.kt
@@ -122,7 +122,11 @@ private fun SqlValuesExpression.argumentType(expression: SqlExpr): IntermediateT
   return when (val parentRule = parent!!) {
     is SqlInsertStmtValues -> {
       val insertStmt = parentRule.parent as SqlInsertStmt
-      insertStmt.columns[argumentIndex].type()
+      if (insertStmt.columnNameList.isNotEmpty()) {
+        insertStmt.columnNameList[argumentIndex].type()
+      } else {
+        insertStmt.columns[argumentIndex].type()
+      }
     }
     is SqlSelectStmt -> {
       val compoundSelect = parentRule.parent as SqlCompoundSelectStmt

--- a/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/queries/MutatorQueryFunctionTest.kt
+++ b/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/queries/MutatorQueryFunctionTest.kt
@@ -370,4 +370,41 @@ class MutatorQueryFunctionTest {
       |}
       |""".trimMargin())
   }
+
+  @Test fun `subexpressions with shared arguments`() {
+    val file = FixtureCompiler.parseSql("""
+      |CREATE TABLE category(
+      |  id TEXT NOT NULL PRIMARY KEY,
+      |  name TEXT NOT NULL,
+      |  description TEXT NOT NULL
+      |);
+      |
+      |save:
+      |INSERT OR REPLACE
+      |INTO category (rowid, id, name, description)
+      |VALUES (COALESCE((SELECT rowid FROM category c2 WHERE id = ?1), NULL), ?1, ?2, ?3);
+      """.trimMargin(), tempFolder)
+
+    val insert = file.namedMutators.first()
+    val generator = MutatorQueryGenerator(insert)
+
+    assertThat(generator.function().toString()).isEqualTo("""
+      |override fun save(
+      |  id: kotlin.String,
+      |  name: kotlin.String,
+      |  description: kotlin.String
+      |) {
+      |  driver.execute(${insert.id}, ""${'"'}
+      |  |INSERT OR REPLACE
+      |  |INTO category (rowid, id, name, description)
+      |  |VALUES (COALESCE((SELECT rowid FROM category c2 WHERE id = ?), NULL), ?, ?, ?)
+      |  ""${'"'}.trimMargin(), 4) {
+      |    bindString(1, id)
+      |    bindString(2, id)
+      |    bindString(3, name)
+      |    bindString(4, description)
+      |  }
+      |}
+      |""".trimMargin())
+  }
 }

